### ISSUE TITLE
Throw driveway deadends

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#4376]
As discussed recently and credit to @kevinkreiser & @dnesbitt61 :

since Amazon adds tons of driveways to OSM and they're almost exclusively private but not tagged that way, our graph gets bloated with pretty useless stuff. One way to deal with it would be to detect them during the graphbuild and just throw them if it' s a single-edge deadend, reducing the graph, improving the performance, yay:)

**Comment by dnesbitt61** [comment]
Could they be removed within the ConstructEdges method? That way any nodes that connect the driveway to the main road would not be added as graph nodes. Removing edges at a later place is less effective (creates "false" nodes that split edges)  unless some sort of edge aggregation logic is created. I've thought about this to remove driveways in the GraphFilter stage but aggregation is difficult.

**Comment by nilsnolde** [comment]
I didn't look into it yet, just wanted to put it in the backlog before it's gone. However, I'm fairly sure we can do it in ConstructEdges, we should have all the necessary info to just remove those nodes/waynodes if we encounter such an edge.

**Comment by kevinkreiser** [comment]
see, just as simple as that :smile:

**Comment by dnesbitt61** [comment]
This includes any driveway marked as:

- highway=service
- service=driveway 
- (with or without access=private)
- Do these need to be dead-ends or can we also remove "semi-circle" driveways (https://www.openstreetmap.org/way/946398713)

What about:

- highway=service, access=private (such as https://www.openstreetmap.org/way/951971091)
- Single edge dead-end that is highway=service only (e.g. https://www.openstreetmap.org/way/951995766)?
- Single edge highway=service connected to same way or different ways?

**Comment by kevinkreiser** [comment]
yeah I agree we could come up with a little function to decide what should be marked as a candidate for removal and that could indeed include other ways. with service ways we might have to take care not to bork places like Japan though that make heavy use of service to just mean low class roads

**Comment by nilsnolde** [comment]
IMO all your highlighted examples would be valid to remove, as they don’t add any topological value, other than waypoint correlation. It’ll be configurable anyways. 

With non-deadends (eg half-circles), do we need to be careful bcs of closures (constructions site or so) on the other non-service way? In that case it might be a valid alternative.
